106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,583 @@
+Everytime an anomaly of the same type is detected we will treat its detection time as the `last_anomaly_detection_time` and update the value of it in the `last_anomaly_detection_time` field of the ConfigMap.
+Now if the anomaly detection happened during the rebalance, the `ConfigMap` will be updated with the anomalies as well as the `last_anomaly_detection_time` will also be updated with last anomaly detection time.
+We will track the rebalance finish time in ConfigMap under the `rebalance_finished_timestamp` field.
+Now to know if the new anomaly was detected during the rebalance we can compare the `last_anomaly_detection_timestamp` and `rebalance_finished_timestamp`.
+If the `last_anomaly_detection_timestamp` was before the `rebalance_finished_timestamp`, that would suggest that the anomaly was detected during the rebalance.
+We can ignore all the anomalies that were detected before the `rebalance_finished_timestamp`.
+This is because it is possible that the new anomalies were fixed during the rebalance.
+Even if the ignored anomalies weren't fixed, the Anomaly Detector can detect them again.
+
+The above hold true for the manual rebalances also. 
+If a user tries a manual rebalance then we will still update the `rebalance_finished_timestamp` to track the last rebalance and avoid auto-rebalancing if the anomaly list is not empty and the `last_anomaly_detection_timestamp` was before the `rebalance_finished_timestamp`.
+
+#### Using other notifiers when using `imbalance` mode
+
+Users cannot configure the notifier if they are utilising the auto-rebalance on imbalanced cluster.
+This is because the operator is using our custom notifier for getting alerts about goal violations. 
+If the users try to override the notifier while the `imbalance` mode is enabled, then the operator would throw warnings in the status field of the Kafka CR and ignore the changes made by the user
+
+```yaml
+status:
+  autoRebalance:
+    state: Idle
+    lastTransitionTime: "2025-09-22T16:04:00Z"
+    modes:
+      - mode: imbalance
+  conditions:
+    - type: Warning
+      status: "True"
+      reason: UnsupportedOperationException
+      message: |
+        Notifier cannot be overridden since `imbalance` mode is enabled in the `auto-rebalance` configuration
+```
+
+If the users really want to have their own way of dealing with the imbalanced clusters then they can disable auto-rebalance in `imbalance` mode and use their own notifier. 
+
+#### What happens if an unfixable goal violation happens
+
+Cruise Control has the ability to figure out if a goal violation can be fixed by the generated optimization proposal or not.
+The goal optimizer class does this job by doing analysis/dry-run if we can fix the anomaly by the generated proposal or not.
+If there is some goal, for example the `DiskDistributionUsage` goal is violated, and we cannot fix it since the all the disks are already completely populated then it is marked as unfixable.
```

Comment:
```suggestion
If a goal such as the `DiskDistributionUsage` goal is violated, but we cannot fix it since the all the disks are already completely populated, then it is marked as unfixable.
```